### PR TITLE
docs: document KNN, PoT, Hybrid models

### PR DESCRIPTION
**Pull request by andrei-stoian-zama**
Not finished, the examples won't work yet, it should be done when the KNN + Hybrid model PRs are merged. 

Closes https://github.com/zama-ai/concrete-ml-internal/issues/3975 
Closes https://github.com/zama-ai/concrete-ml-internal/issues/3976 
Closes https://github.com/zama-ai/concrete-ml-internal/issues/3977

### CODE REVIEW

File: docs/built-in-models/nearest-neighbors.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+|        Concrete ML         | scikit-learn                                                                                                          |
+| :------------------------: | --------------------------------------------------------------------------------------------------------------------- |
+| [KNeighborsClassifier](<>) | [KNeighborsClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html) |
+
+## Example usage
+
+```python
+from concrete.ml.sklearn import KNeighborsClassifier
+
+concrete_classifier = KNeighborsClassifier(n_bits=2,     n_neighbors=3)
```
Comment (kcelia) [inline docs/built-in-models/nearest-neighbors.md:14]: too much space

Comment (RomanBredehoft) [inline docs/built-in-models/nearest-neighbors.md:14]: weird spaces no ?

File: docs/built-in-models/nearest-neighbors.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+```
+
+The `KNeighborsClassifier` class quantizes the training dataset that is given to `.fit` with the specified number of bits, `n_bits`. As this value must be kept low to comply with [accumulator size constraints](../getting-started/concepts.md#model-accuracy-considerations-under-fhe-constraints) the accuracy of the model will depend heavily a well-chosen value `n_bits` and the dimensionality of the data.
+
+The FHE inference latency of this model is heavily influenced by the `n_bits`, the dimensionality of the data. Furthermore, the size of the dataset has a linear impact on the complexity of the data and the number of nearest neighbors, `n_neighbors`, also plays a role.
+
+The KNN computation executes in FHE in $$O(Nlog^2k)$$ steps, where $$N$$ is the training dataset size and $$k$$ is `n_neighbors`. Each step requires several PBS of the precision required to represent the distances between test vectors and the training dataset.
```
Comment (RomanBredehoft) [inline docs/built-in-models/nearest-neighbors.md:21]: on github the "$$ $$" are centered on a different line, is that expected ?

File: docs/built-in-models/neural-networks.md [hunk 3]
```diff
@@ -17,17 +17,14 @@ While `NeuralNetClassifier` and `NeuralNetClassifier` provide scikit-learn-like
 Good quantization parameter values are critical to make models [respect FHE constraints](../getting-started/concepts.md#model-accuracy-considerations-under-fhe-constraints). Weights and activations should be quantized to low precision (e.g., 2-4 bits). The sparsity of the network can be tuned [as described below](neural-networks.md#overflow-errors) to avoid accumulator overflow.
 {% endhint %}
 
+{% hint style="warning" %}
+Using `nn.ReLU` as the activation function benefits from an optimization where quantization uses powers-of-two scales. This results in much faster inference times in FHE, thanks to a TFHE primitive that performs fast division by powers of two.
```
Comment (RomanBredehoft) [inline docs/built-in-models/neural-networks.md:21]: do we have a documentation to link about it ?

File: docs/built-in-models/neural-networks.md [hunk 4]
```diff
@@ -36,11 +33,6 @@ n_inputs = 10
 n_outputs = 2
 params = {
     "module__n_layers": 2,
-    "module__n_w_bits": 2,
```
Comment (RomanBredehoft) [inline docs/built-in-models/neural-networks.md:39]: why the removal ?

**Review by RomanBredehoft — DISMISSED** [review]
a few questions, but apart from that all good for me

File: docs/built-in-models/neural-networks.md [hunk 5]
```diff
@@ -17,17 +17,14 @@ While `NeuralNetClassifier` and `NeuralNetClassifier` provide scikit-learn-like
 Good quantization parameter values are critical to make models [respect FHE constraints](../getting-started/concepts.md#model-accuracy-considerations-under-fhe-constraints). Weights and activations should be quantized to low precision (e.g., 2-4 bits). The sparsity of the network can be tuned [as described below](neural-networks.md#overflow-errors) to avoid accumulator overflow.
 {% endhint %}
 
+{% hint style="warning" %}
+Using `nn.ReLU` as the activation function benefits from an optimization where quantization uses powers-of-two scales. This results in much faster inference times in FHE, thanks to a TFHE primitive that performs fast division by powers of two.
+{% endhint %}
```
Comment (kcelia) [inline docs/built-in-models/neural-networks.md:22]: Maybe we can add the link to this primitive.

File: docs/built-in-models/nearest-neighbors.md (hunk 2 above)
Comment (kcelia) [inline docs/built-in-models/nearest-neighbors.md:21]: I suggest:

The concrete KNN algorithm operates in  O(Nlog⁡2k) steps, where N .....

Each step needs several PBS of the precision required to represent the distances between query vector and the training dataset.

Comment (kcelia) [reply docs/built-in-models/nearest-neighbors.md:21]: by the way, I find  "Each step requires several PBS of the precision required", a bit unclear. 

So maybe that way: 

Each step requires several PBSs to effectively represent the precision of distances between query vector and the training dataset.

We don't talk about the computation of the labels for the k-nearest neighbors ?

File: docs/built-in-models/nearest-neighbors.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+| [KNeighborsClassifier](<>) | [KNeighborsClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html) |
+
+## Example usage
+
+```python
+from concrete.ml.sklearn import KNeighborsClassifier
+
+concrete_classifier = KNeighborsClassifier(n_bits=2,     n_neighbors=3)
+```
+
```
Comment (kcelia) [inline docs/built-in-models/nearest-neighbors.md:16]: I suggest:

The `KNeighborsClassifier` class quantizes quantization on the training data-set provided to the .fit method, with the specified number of bits, `n_bits`. It is crucial to maintain a low value for n_bits to comply to the accumulator size constraints and the algorithm's time complexity, which is closely tied to three key factors: n_bits, data dimensionality, and the size of the training dataset.

The Concrete K-nearest neighbors computation operates in O(Nlog⁡2k) steps, where $$N$$ is the training dataset size and $$k$$ is `n_neighbors`. Furthermore, the size of the dataset has a linear impact on the complexity of the data and the number of nearest neighbors, `n_neighbors`, also plays a role.

File: docs/built-in-models/neural-networks.md [hunk 7]
```diff
@@ -89,4 +82,6 @@ You can give weights to each class to use in training. Note that this must be su
 
 ### Overflow errors
 
-The `n_hidden_neurons_multiplier` parameter influences training accuracy as it controls the number of non-zero neurons that are allowed in each layer. Increasing `n_hidden_neurons_multiplier` improves accuracy, but should take into account precision limitations to avoid an overflow in the accumulator. The default value is a good compromise that avoids an overflow in most cases, but you may want to change the value of this parameter to reduce the breadth of the network if you have overflow errors. A value of 1 should be completely safe with respect to overflow.
+The `n_accum_bits` parameter influences training accuracy as it controls the number of non-zero neurons that are allowed in each layer. Increasing `n_accum_bits` improves accuracy, but should take into account precision limitations to avoid an overflow in the accumulator. The default value is a good compromise that avoids an overflow in most cases, but you may want to change the value of this parameter to reduce the breadth of the network if you have overflow errors.
```
Comment (kcelia) [inline docs/built-in-models/neural-networks.md:85]: Good catch

File: docs/advanced-topics/advanced_features.md [hunk 8]
```diff
@@ -187,7 +187,7 @@ In practice, the process looks like this:
 1. Update P = P - 1
 1. repeat steps 2 and 3 until the accuracy loss is above a certain, acceptable threshold.
 
-An example of such implementation is available in [evaluate_torch_cml.py](../../use_case_examples/cifar_brevitas_training/evaluate_one_example_fhe.py) and [CifarInFheWithSmallerAccumulators.ipynb](../../use_case_examples/cifar_brevitas_finetuning/CifarInFheWithSmallerAccumulators.ipynb)
+An example of such implementation is available in [evaluate_torch_cml.py](../../use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py) and [CifarInFheWithSmallerAccumulators.ipynb](../../use_case_examples/cifar/cifar_brevitas_finetuning/CifarInFheWithSmallerAccumulators.ipynb)
```
Comment (jfrery) [inline docs/advanced-topics/advanced_features.md:191]: Thanks!

File: docs/advanced-topics/hybrid-models.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
+# Hybrid model deployment
+
+FHE enables cloud applications to process private user data without running the risk of data leaks. Furthermore, deploying ML models in the cloud is advantageous as it eases model updates, allows to scale to large numbers of users by using large amounts of compute power, and protects model IP by keeping the model on a trusted server instead of the client device.
+
+However, not all applications can be easily converted to FHE computation and the computation cost of FHE may make a full conversion exceed latency requirements.
+
```
Comment (jfrery) [inline docs/advanced-topics/hybrid-models.md:6]: I find the sentence a bit hard to follow. I propose


```suggestion
For certain applications, transitioning to FHE computation may not be straightforward due to inherent complexities. Additionally, the computational overhead of FHE might cause latency issues that surpass acceptable thresholds.
```

File: docs/advanced-topics/hybrid-models.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+Hybrid models are a compromise between on-premise or on-device deployment and full cloud deployment. Hybrid deployment means parts of the model are executed on the client side and parts are executed in FHE on the server side. Concrete ML supports hybrid deployment of neural network models such as MLP, CNN and Large Language-Models.
+
```
Comment (jfrery) [inline docs/advanced-topics/hybrid-models.md:8]: I feel like on-prem and on-device is the same? I propose this:


```suggestion
Hybrid models provide an effective balance between on-device deployment and cloud-based deployment. This approach entails executing parts of the model directly on the client side, while other parts are securely processed with FHE (Fully Homomorphic Encryption) on the server side. Concrete-ML facilitates the hybrid deployment of various neural network models, including but not limited to, MLP (Multilayer Perceptron), CNN (Convolutional Neural Network), and Large Language Models.
```

File: docs/advanced-topics/hybrid-models.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+{% hint style="warning" %}
+If model IP protection is important, care must be taken in choosing the parts of a model to be executed on the cloud. Some
+black-box model stealing attacks rely on knowledge distillation
+or on differential methods. As a general rule, the difficulty
+to steal a machine learning model is proportional to the size of the model, in terms of numbers of parameters and model depth.
+{% endhint %}
```
Comment (jfrery) [inline docs/advanced-topics/hybrid-models.md:14]: I am not sure what users should do with that information. Should we point them to the litterature maybe?

File: docs/advanced-topics/hybrid-models.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    inputs,
+    n_bits=8,
+)
+
+
+models_dir = Path(os.path.abspath('')) / "compiled_models"
+models_dir.mkdir(exist_ok=True)
+model_dir = models_dir / model_name
+
+```
```
Comment (jfrery) [inline docs/advanced-topics/hybrid-models.md:64]: I think we can get rid of all the path handling + tuple_to_underscore_str for the codeblock?

File: docs/advanced-topics/hybrid-models.md [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+```python
+hybrid_model.save_and_clear_private_info(model_dir, via_mlir=True)
+```
+
+## Server Side Deployment
+
+The [`save_and_clear_private_info`](../developer-guide/api/concrete.ml.torch.hybrid_model.md#method-save_and_clear_private_info) function serializes the FHE circuits
+corresponding to the various parts of the model that were chosen to be moved
+server-side. Furthermore it saves all necessary information required
+to serve these sub-models with FHE, using the [`FHEModelDev`](../developer-guide/api/concrete.ml.deployment.fhe_client_server.md#class-fhemodeldev) class.
```
Comment (jfrery) [inline docs/advanced-topics/hybrid-models.md:77]: Also something important is that the save_and_clear_private_info removed all the IP related weights.

File: docs/built-in-models/neural-networks.md [hunk 14]
```diff
@@ -56,13 +48,14 @@ The figure above right shows the Concrete ML neural network, trained with Quanti
 ### Architecture parameters
 
 - `module__n_layers`: number of layers in the FCNN, must be at least 1. Note that this is the total number of layers. For a single, hidden layer NN model, set `module__n_layers=2`
-- `module__activation_function`: can be one of the Torch activations (e.g., nn.ReLU, see the full list [here](../deep-learning/torch_support.md#activations))
+- `module__activation_function`: can be one of the Torch activations (e.g., nn.ReLU, see the full list [here](../deep-learning/torch_support.md#activations)). Neural networks with `nn.ReLU` activation benefit from specific optimizations that make them around 10x faster than networks with other activation functions.
```
Comment (jfrery) [inline docs/built-in-models/neural-networks.md:51]: Great!

File: docs/built-in-models/neural-networks.md [hunk 15]
```diff
@@ -56,13 +48,14 @@ The figure above right shows the Concrete ML neural network, trained with Quanti
...
 
 ### Quantization parameters
 
 - `n_w_bits` (default 3): number of bits for weights
 - `n_a_bits` (default 3): number of bits for activations and inputs
-- `n_accum_bits` (default 8): maximum accumulator bit-width that is desired. The implementation will attempt to keep accumulators under this bit-width through [pruning](../advanced-topics/pruning.md) (i.e., setting some weights to zero)
+- `n_accum_bits`: maximum accumulator bit-width that is desired. By default, this is unbounded, which, for weight and activation bitwidth settings, [may make the trained networks fail in compilation](#overflow-errors). When used, the implementation will attempt to keep accumulators under this bit-width through [pruning](../advanced-topics/pruning.md) (i.e., setting some weights to zero)
+- `power_of_two_scaling`: forces quantization scales to be powers-of-two, which, when coupled with the ReLU activation, benefits from strong FHE inference time optimization
```
[the remark this answers is not in the document]
Comment (jfrery) [reply docs/built-in-models/neural-networks.md:58]: If we don't have something on quantization.md about this type of quantization maybe it's worth to point to a paper or another source?

File: docs/getting-started/showcase.md [hunk 16]
```diff
@@ -61,7 +61,7 @@ Simpler tutorials that discuss only model usage and compilation are also availab
          <td></td>
          <!--- start -->
          <td><a href="../.gitbook/assets/demo_nn_splitting.png">client-server-1.png</a></td>
-         <td><a href="../../use_case_examples/cifar_brevitas_with_model_splitting">use_case_examples/cifar_brevitas_with_model_splitting</a></td>
+         <td><a href="../../use_case_examples/cifar/cifar_brevitas_with_model_splitting">use_case_examples/cifar_brevitas_with_model_splitting</a></td>
          <!--- end -->
```
Comment (jfrery) [inline docs/getting-started/showcase.md:65]: Hmm weird I should have done that I guess my bad.

File: README.md [hunk 17]
```diff
@@ -132,11 +132,11 @@ Various tutorials are proposed for the [built-in models](docs/built-in-models/ml
...
 
-- [FHE neural network splitting for client/server deployment](use_case_examples/cifar_brevitas_with_model_splitting): we explain how to split a computationally-intensive neural network model in two parts. First, we execute the first part on the client side in the clear, and the output of this step is encrypted. Next, to complete the computation, the second part of the model is evaluated with FHE. This tutorial also shows the impact of FHE speed/accuracy trade-off on CIFAR10, limiting PBS to 8-bit, and thus achieving 62% accuracy.
+- [FHE neural network splitting for client/server deployment](use_case_examples/cifar/cifar_brevitas_with_model_splitting): we explain how to split a computationally-intensive neural network model in two parts. First, we execute the first part on the client side in the clear, and the output of this step is encrypted. Next, to complete the computation, the second part of the model is evaluated with FHE. This tutorial also shows the impact of FHE speed/accuracy trade-off on CIFAR10, limiting PBS to 8-bit, and thus achieving 62% accuracy.
```
Comment (jfrery) [inline README.md:139]: Thanks!

**Review by jfrery — COMMENTED** [review]
Thanks for handling this! I have a few comments.

File: docs/built-in-models/nearest-neighbors.md (hunk 2 above)
Comment (andrei-stoian-zama) [reply docs/built-in-models/nearest-neighbors.md]: we need the double dollars to make latex work in our docs, on github it has always been a bit broken

File: docs/built-in-models/neural-networks.md (hunk 4 above)
Comment (andrei-stoian-zama) [reply docs/built-in-models/neural-networks.md:39]: if default parameters are good, it's best to give examples with default parameters -> it's best to showcase ease-of-use instead of complexity

File: docs/built-in-models/neural-networks.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply docs/built-in-models/neural-networks.md:58]: I don't think the user cares much about what this kind of quantization does. They care that if they turn it off the models are slower.  But you're right, in advanced topics we can discuss it.

**Comment by fd0r** [comment]
Needs to be rebased

File: docs/built-in-models/neural-networks.md (hunk 3 above)
Comment (andrei-stoian-zama) [reply docs/built-in-models/neural-networks.md:21]: added one

File: docs/built-in-models/neural-networks.md (hunk 5 above)
Comment (andrei-stoian-zama) [reply docs/built-in-models/neural-networks.md:22]: yes, adding it

File: docs/advanced-topics/hybrid-models.md (hunk 11 above)
Comment (andrei-stoian-zama) [reply docs/advanced-topics/hybrid-models.md:14]: I wanted a "disclaimer" - simply making a model hybrid doesn't fix all problems.

File: docs/advanced-topics/hybrid-models.md (hunk 12 above)
Comment (andrei-stoian-zama) [reply docs/advanced-topics/hybrid-models.md:64]: no, not possible actually, the exported paths are fixed in CML code

File: docs/advanced-topics/hybrid-models.md (hunk 13 above)
Comment (andrei-stoian-zama) [reply docs/advanced-topics/hybrid-models.md:77]: ok, I'll add that

File: docs/built-in-models/neural-networks.md (hunk 4 above)
Comment (RomanBredehoft) [reply docs/built-in-models/neural-networks.md:39]: yeah indeed !

File: docs/built-in-models/nearest-neighbors.md (hunk 2 above)
Comment (RomanBredehoft) [reply docs/built-in-models/nearest-neighbors.md]: got it

File: docs/built-in-models/neural-networks.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (RomanBredehoft) [reply docs/built-in-models/neural-networks.md]: great !

File: docs/advanced-topics/hybrid-models.md (hunk 9 above)
Comment (RomanBredehoft) [reply docs/advanced-topics/hybrid-models.md:6]: agree !

File: docs/advanced-topics/hybrid-models.md (hunk 11 above)
Comment (RomanBredehoft) [reply docs/advanced-topics/hybrid-models.md:14]: I would maybe link to some literature on KD attacks indeed !

File: docs/advanced-topics/hybrid-models.md [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+# Create a hybrid model
+hybrid_model = HybridFHEModel(model, [submodule_name])
+hybrid_model.compile_model(
+    inputs,
+    n_bits=8,
+)
+
+
+models_dir = Path(os.path.abspath('')) / "compiled_models"
```
Comment (RomanBredehoft) [inline docs/advanced-topics/hybrid-models.md:60]: `Path(os.path.abspath(''))` is identical to `Path('.')>resolve()` right ? (avoid importing `os`)

File: docs/advanced-topics/hybrid-models.md (hunk 12 above)
Comment (jfrery) [reply docs/advanced-topics/hybrid-models.md:64]: Damn that's not pretty in the doc. Alright

File: docs/advanced-topics/hybrid-models.md (hunk 11 above)
Comment (fd0r) [reply docs/advanced-topics/hybrid-models.md:14]: Maybe worth creating an issue and properly documenting this in a second time?

**Review by fd0r — APPROVED** [review]
Looks good to me!

File: docs/advanced-topics/hybrid-models.md (hunk 18 above)
Comment (andrei-stoian-zama) [reply docs/advanced-topics/hybrid-models.md:60]: I like `os.path` :)
